### ISSUE TITLE
Problem in recreating results

### ISSUE THREAD

**Issue by prashastbindal** [DuaneNielsen/keypoints#1]
Hi DuaneNielsen,

First of all thank you for doing this implementation, it's a great help to us! I recently tried recreating the results on Celeb A dataset, but I am unable to achieve results as good as the original research paper. Were you able to recreate the original paper results with your training? If so, would it be possible for you to share a pretrained model that you might have?

I am sharing a few screenshots from my run

![WhatsApp Image 2019-12-02 at 18 21 25](https://user-images.githubusercontent.com/13622022/70004040-fe774980-1532-11ea-93a1-ad7b1ef56e16.jpeg)

![WhatsApp Image 2019-12-02 at 18 22 00](https://user-images.githubusercontent.com/13622022/70005065-477ccd00-1536-11ea-8e82-e1fb0b94d300.jpeg)

**Comment by DuaneNielsen** [comment]
Hey prashastbindal,

Thanks for your comment, glad to see you share my enthusiasm about this paper!

I'm still actively developing the project at the moment.  

I have been unable to reproduce the results, although I fully intend to do so.

I downloaded and ran Thomas's original code on the weekend.  https://github.com/tomasjakab/imm

A few things I noticed...

1. He runs the training loop for 3 M batches!  This takes 10 days on my RTX2070
2. His thin plate spline distortions are a bit lower than what I'm using.
3. He uses Adam lr1e-4, I've been mostly using SGD, but I'm changing to Adam
4. He uses perceptual loss, I'm using MSE reconstruction loss.  I don't think this is a big deal due to https://arxiv.org/abs/1906.11883 which also uses reconstruction loss. 
5. He includes a loss mask calculated from the tps distortion, which I have not implemented.  I think this is the most likely cause of the failures at the moment.

At some point, I want to verify that my math is the same as his.

Given that I can't wait 10 days for a result!  I'm going to try using atari pong images/transitions instead in a very simplified network to see if I can get that working first.

If you are interested in collaborating, I'd be more than open, we can use this issue to communicate.

Cheers

Duane

**Comment by prashastbindal** [comment]
Hi DuaneNielsen,

Thanks for your detailed reply earlier. I found this and so I thought of sharing this with you. The authors of imm say that 100-150k iterations might be just enough for the results : 

https://github.com/tomasjakab/imm/issues/1

Regards
Prashast

**Comment by DuaneNielsen** [comment]
Thanks!  That is incredibly helpful.  I'll try it out.

I'm verifying the correctness of my bottleneck at the moment also.

When I looked into it, is suspect batch normalizing the output of the heatmap keypoint network might be the problem.

Reason being, numbers in the range 0-1 in the heatmap when squashed with softmax, generate a distribution that has very little variance.

So I'm still hunting for the exact issue!

**Comment by DuaneNielsen** [comment]
Ok, here are a few sceenshots from my latest run.  It's still going, but as you can see, there is marked improvement...

![Selection_001](https://user-images.githubusercontent.com/11070508/70870120-60c54680-1f44-11ea-9c83-ed8b410ebd45.png)
![Selection_002](https://user-images.githubusercontent.com/11070508/70870122-6589fa80-1f44-11ea-8730-248bf5c45b99.png)
![Selection_003](https://user-images.githubusercontent.com/11070508/70870124-6884eb00-1f44-11ea-8af0-3006b891f637.png)

The loss mask was critical.  Without it, the keypoint network learns the TPS transform by detecting the shape of the edges of the image.

The other problem was I was feeding batch normalized input to a softmax.  Always a bad idea, because softmax is actually a type of normalization.  So I was normalizing a normalized output.

Switching to Adam from SGD was also very helpful, learning proceeded at a faster rate.

Finally, knowing how long to run the training for was very helpful.. I'm getting results in the described 150k minibatches.  Actually, you can see keypoints forming on the eyebrows very early in training.

I also reduced the amount of TPS and Rotation noise.

The command I ran to get these results was 

```
python3 keypoints.py --model_name keynet_vgg_11 --run_id 1 --dataset celeba --batch_size 30 --num_keypoints 10 --epochs 24 --opt_level O2 --run_type full
```

tag here : https://github.com/DuaneNielsen/keypoints/releases/tag/v0.1

I required the --opt_level O2 to run in mixed precision on the RTX2070 due to full precision not being able to fit the model on device with batch size 30.  Mixed precision doesn't appear to affect the results.  There are occasional gradient discards, but nothing major.

The run took a little under 24 hours to get through 24 epochs on the RTX2070.  So not speedy, but bearable.

**Comment by DuaneNielsen** [comment]
Here is a panel from 150k+ mini-batches of training.

![image](https://user-images.githubusercontent.com/11070508/71146128-df3a1680-21d8-11ea-84cc-a120577e63ca.png)

The test/train loss is looking very nice, so it could be trained further.

![image](https://user-images.githubusercontent.com/11070508/71146256-638c9980-21d9-11ea-8724-67086b8f5485.png)

On to my 3rd run now.

**Comment by DuaneNielsen** [comment]
Ok, I am able to reproduce the results consistently now...

Changing to use explicit padding modules instead of Conv2d's "padding" flag, and filling using the "border" feature on the padding module.  I'm now convinced that the Conv2d padding flag is evil!

The transporter network is numerically unstable, but even so, it does a pretty good job.
